### ISSUE TITLE
Improvements to TCPServer#accept

### ISSUE THREAD

**Issue by ioquatix** [socketry/async#11]
I've been thinking about how we can improve this API.

I like the idea of something like `TCPServer.accept_each do |peer| ....`

Wondering what your though might be @tarcieri 

I think it's more canonical ruby than the pythonic `task.with(server.accept) do |peer| ... end while true`

**Comment by tarcieri** [comment]
You want to monkeypatch `TCPServer`?

**Comment by picatz** [comment]
@tarcieri Is it wrong of me to like this idea?

**Comment by tarcieri** [comment]
I have generally grown to loathe core extensions. My advice before monkeypatching core is to seriously consider if the problem would be better solved by object composition.

If you really want to monkeypatch core, it's probably worth opening an issue on https://bugs.ruby-lang.org and discuss it there. That's not to say you should wait for core to accept your idea, just generally that core extensions should be the sorts of things you'd like to make it upstream eventually.

That said, as a purely additive change, and one that seems fairly decoupled from everything else, this does look kind of neat.

**Comment by picatz** [comment]
I appreciate the explanation, and I think that makes a lot of sense.

_We should totes do it anyway though amiright?_

**Comment by ioquatix** [comment]
It's not a monkey patch it sits within the TCPServer wrapper

**Comment by tarcieri** [comment]
A wrapper is a perfectly fine place to put it

**Comment by picatz** [comment]
![obligatory picat image](https://cloud.githubusercontent.com/assets/14850816/25064805/3a6a5e40-21d0-11e7-8ab6-799eabef2006.jpg)
_me chilln hur like_

**Comment by ioquatix** [comment]
Okay, so I thought about this a lot.

@tarcieri you are right, there are fundamental issues with `TCPServer`, `TCPSocket`, and `UDPSocket`. Not to mention, that some of the underlying code of those classes is buggy: https://bugs.ruby-lang.org/issues/13409

So, I revisited `Async::Socket`. I've improved it to the point where it can be easily used for all networking purposes.

I've added `Async::Socket.bind` which binds to a given `sockaddr`, `Async::Socket.accept` which binds and then calls accept in a loop.

The other wrappers for TCP and UDP still work. I may even improve them.. but for now, it makes most sense to use `Socket` itself.
